lib/hopper.rb

```diff
@@ -61,8 +88,32 @@ def clear
       @registrations = []
     end
 
+    def semaphore
+      @semaphore ||= Mutex.new
+    end
+
     private
 
+    def initialize_hopper
```

Comment:
You may have already considered this, so this more is for my curiosity: why not just make this method atomic?
```
def initialize_hopper
  semaphore.synchronize do
    @state = INITIALIZING
    ...
    @state = INITIALIZED
  end
ensure
  semaphore.unlock
end
```
It may help reduce locking the semaphore in other places, hopefully making it more clear which parts of the code represent critical sections

Reply:
I needed to put the state ```INITIALIZING``` only when it's called from the initializer ```init_channel```. This is to differentiate between cases where rabbitmq is not even configured on the system from the cases where is configured but it failed initialization. 

When a message is published and the service is not ```INITIALIZED``` it will attempt to initialize only if it's in ```INITIALIZING``` state

Reply:
That's fine - I'm more concerned that `initialize_hopper` seems to be implicitly a critical section. If someone needed to add a codepath that calls `initialize_hopper`, they would need to manually add thread safety themselves. Is it possible to move the lock into the `initialize_hopper` method itself?

Reply:
it is private, it should not be invoked directly.

Reply:
Yes, that's true. It can't be called by external sources. I'm thinking more in the case where someone is adding codepaths to `hopper.rb` directly.

I should mention that what I am suggesting is stylistic more than functional, so the net gain may not be worth the change. If you don't think it's necessary, that's fine too